### PR TITLE
Add Blackfire article

### PR DESCRIPTION
**Pull request by Anahkiasen**

### CODE REVIEW

File: _posts/2015-10-30-blackfire.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+title: Set Blackfire to the rain
+date: 2015-10-30T00:00:00.000Z
+author: maxime
+excerpt: 'Discover Blackfire, an awesome performance profiling tool'
+---
+
+# Set Blackfire to the rain
+If you've ever worked on any PHP application, or package, or _anything_ you know that debugging performance issues is **hard**. There are several ways to ease the pain a little: debug bars, putting timers a bit everywhere. Or if you're courageous like I was for a long time, you use `xdebug` snapshots which requires you to configure it, and then parsing the snapshots take a long time, etc.
+
+Recently Blackfire came out of beta, it's a tool aimed at making all those a thing of the past by providing a modern powerful performance analysis tool, that is both quick and easy to use. When I discovered it during its first public beta I was like a kid on a toys store. Where all the toys would be made of candy.<br>Today I want to make you discover Blackfire if you don't already know it because I feel like it's not as known as it should be.
```
Comment (tonysm) [inline _posts/2015-10-30-blackfire.md]: maybe ? `it's a tool aimed at making all those profiling options a thing of the past`

Comment (tonysm) [inline _posts/2015-10-30-blackfire.md]: also `I was like a kid in a toy store`?

**Comment by tonysm** [comment]
Nice one! :+1:

File: _posts/2015-11-06-blackfire.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
+---
+layout: post
+title: Set Blackfire to the rain
+date: 2015-11-06T00:00:00.000Z
+author: maxime
+excerpt: 'Discover Blackfire, an awesome performance profiling tool'
+---
+
+# Set Blackfire to the rain
+If you've ever worked on any PHP application, or package, or _anything_ you know that debugging performance issues is **hard**. There are several ways to ease the pain a little: debug bars, putting timers a bit everywhere. Or if you're courageous like I was for a long time, you use `xdebug` snapshots which requires you to configure it, and then parsing the snapshots take a long time, etc.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: > you use `xdebug` snapshots which requires you to configure it, and then parsing the snapshots take a long time, etc.

you use `xdebug` snapshots which requires you to configure it, and then take snapshots, which takes a long time, then parse the snapshots, etc.

Was it meant like this ^?

File: _posts/2015-11-06-blackfire.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+Recently Blackfire came out of beta, it's a tool aimed at making all these profiling options a thing of the past by providing a modern powerful performance analysis tool, that is both quick and easy to use. When I discovered it during its first public beta I was like a kid in a toy store. Where all the toys would be made of candy.
+Today I want to make you discover Blackfire if you don't already know it because I feel like it's not as known as it should be.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: know it**,** because

File: _posts/2015-11-06-blackfire.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+## Blackfire
+![](http://i.imgur.com/vRvy2Wd.png)
+
+Blackfire is a tool created by [Sensiolabs](https://sensiolabs.com/) – which you may know from this little unknown framework called [Symfony](http://symfony.com/) or from [Sensiolabs Insights](https://insight.sensiolabs.com/). It's tailored for doing advanced performance profiling of PHP applications – be it web applications, CLI tools, packages, etc. Anything really. And it is _amazing_.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: Sensio**L**abs

File: _posts/2015-11-06-blackfire.md (hunk 2 above)
Comment (Anahkiasen) [inline _posts/2015-11-06-blackfire.md]: No it's the actual parsing of the snapshots that takes a while. I mean taking the snapshots is done during the task you're profiling and it does slow it down a bit but mostly it's importing the snapshots in PHPStorm that is slow, took around 1mn per snapshot when I was working on Rocketeer

Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: Alright, then:

and then parsing the snapshots **which** take**s** a long time

File: _posts/2015-11-06-blackfire.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+brew install blackfire-agent
+
+sudo blackfire-agent -register
+```
+
+And that's it. Even better: want to install Blackfire from a service like Forge, Fortrabbit or Heroku? It's integrated right into them, just need to fill a form and you're good to go:
+
+![](http://i.imgur.com/S4w0IU2.png)
+
+You can see more informations [in the Getting Started](https://blackfire.io/getting-started) of the documentation.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: more information~~s~~

File: _posts/2015-11-06-blackfire.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+## Profiling web applications
+Now that we have Blackfire, what can we do with it? Let's try to profile the performances of a PHP web application. It can be any application, **you do not need to add any code to it**, Blackfire just has to be installed on the server that runs it and it'll hook itself into PHP to gather statistics.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: you can stop the sentence after "you do not need to add any code to it"

File: _posts/2015-11-06-blackfire.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+The easiest way to debug an application on a server where Blackfire is present is to get [the Chrome extension](https://chrome.google.com/webstore/detail/blackfire-companion/miefikpgahefdbcgoiicnmpbeeomffld?hl=en) (or Lynx or whatever). Then go on a webpage, click on the Blackfire icon, and then on the big red "Profile" button. You will see a bar appear at the top of your page, it'll run for a very short while and then give you informations about the current request:
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: information~~s~~

File: _posts/2015-11-06-blackfire.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+### Understanding Blackfire profiles
+The screen is split into three parts:
+- The toolbar
+- The function list
+- The call graph
+
+The most important one is the **call graph**: it is a visual representation of all the methods that were called during the current request, and which of them took up most of the time. Blackfire will highlight which _paths_ in your application are also the most time-consuming ones, those are called _hot paths_.
+
+You'll also see **nodes** with a colored background, those are the most intensive nodes of your request. If you click on one you'll see it highlighted with more informations on the part on the left of the screen: the function list. It'll tell you, in general during this request, how much _exclusive_ time was spent in this method – eg. without considering calls this method might have made to other methods. Just, the time it took _by itself_. You can switch that to _inclusive_ time by clicking on the headers at the top of the list of methods.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: information~~s~~

File: _posts/2015-11-06-blackfire.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+If you're afraid your one test run is not representative and would like to get an "average" view of performances you can also use the `--samples` flag to execute something X times and create a profile from the average metrics of those X times:
+
+```bash
+blackfire run --samples="10" artisan some:scheduled-job
+```
+
+Would run our job 10 times per example.
+
+### References and comparisons
+Now once you have your first profile and you start working on it, most likely you'll want to more precisely the impact the changes you do on the codebase had on the performances. That's where references come into play: they're special profiles that serve as "reference point". You can create one by picking "Create new reference" from the Chrome extension dropdown or passing the `--new-reference` flag in the CLI.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: > Now once you have your first profile and you start working on it, most likely you'll want to more precisely the impact the changes you do on the codebase had on the performances.

Now once you have your first profile and you start working on it, most likely you'll want to **know** more precisely the impact the changes you do on the codebase had on the performances.

Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: it's a hard sentence to read.

File: _posts/2015-11-06-blackfire.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+Once you do another profile, you'll be offered the possibility to compare to that reference from the dropdown (or through the `--reference={reference-id}` flag). If you do that you'll see a slightly different profile screen which compares every metric to the reference profile:
+
+![](http://i.imgur.com/uLJQRTc.png)
+
+This gives you the ability to measure your progress as you make changes and see if you're going into the right direction or not. Note that you can also compare quickly two profiles (any profile) on your Blackfire dashboard:
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: also compare ~~quickly~~ two profiles **quickly**

File: _posts/2015-11-06-blackfire.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+![](http://i.imgur.com/G9I44Cr.png)
+
+## Profiling APIs and web services
+You can also profile any URL or API by using the `blackfire curl` command. Options are the same as for `blackfire run` so you can use samples and things like that. Note that the server has to have the Blackfire PHP extension, same as for web applications (since principle is the same anyway):
+
+```bash
+blackfire curl http://my-website/api/some-call
+```
+
+You can even straight up copy curl requests from the Chrome devtools and pass them to Blackfire.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: screenshot? many people don't know how to do this in chrome... nice little tip to add

File: _posts/2015-11-06-blackfire.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+Once you have profiles, you can share them to anybody via the little share icon at the top right corner of any profile. You can even embed those profiles into any page via an iframe:
+
+<iframe frameborder="0" width="100%" height="400" allowfullscreen src="https://blackfire.io/profiles/a9ebc73e-66bc-48a4-9d3a-9b9dbaf534da/embed"></iframe>
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: this profile "has been removed by the user"?

Comment (Anahkiasen) [inline _posts/2015-11-06-blackfire.md]: Oops.

File: _posts/2015-11-06-blackfire.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+  "Homepage should not hit the DB":
+    path: "/"                                 # only apply the assertions for the homepage
+    assertions:
+      - "metrics.sql.queries.count == 0"      # no SQL statements executed
+      - "main.peak_memory < 10mb"             # memory does not exceed 10mb
+      - "metrics.output.network_out < 100kb"  # the response size is less than 100kb
+  "My queue job should be fast enough":
+    command: "artisan some:scheduled-job"
+    assertions:
+      - "main.wall_time < 10s" # wall clock time is less than 2s
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: > - "main.wall_time < 10s" # wall clock time is less than 2s

10s? or 2s?

File: _posts/2015-11-06-blackfire.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+```
+
+Now every time you run Blackfire it'll add a new tab on your profiles showing you if your profile failed those tests or not. And you can test any metrics (memory, CPU, time) and you can also run comparisons against references and all kinds of things – it's a really powerful feature.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: maybe add a reference link for this paragraph: https://blackfire.io/doc/performance-tests

Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: > **And** you can test... **and** you can also...

**Y**ou can test

File: _posts/2015-11-06-blackfire.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+```bash
+# Before
+artisan some:scheduled-job
+
+# After
+blackfire run artisan some:scheduled-job
+```
+
+Your command will run and Blackfire will give you a link to the profile and some informations: time it took, number of SQL requests, etc.
```
Comment (hannesvdvreken) [inline _posts/2015-11-06-blackfire.md]: information~~s~~

File: _posts/2015-11-06-blackfire.md (hunk 14 above)
Comment (Anahkiasen) [inline _posts/2015-11-06-blackfire.md]: YOUUU CAN TEEEST. YOUU CAN JIIIIVE. HAVIIING THE TIIIME OF YOUUR LIIIIFE

**Comment by hannesvdvreken** [comment]
:+1:

File: _posts/2015-11-06-blackfire.md (hunk 14 above)
Comment (tonysm) [inline _posts/2015-11-06-blackfire.md]: :stuck_out_tongue:

**Comment by hannesvdvreken** [comment]
:ship:

**Pull request merged**